design/pipeline-artifacts.md

```diff
@@ -128,22 +133,22 @@ You can add files to both the default artifact and to named artifacts until the
       TODO-some-cool-deploy-script-here $(Pipelines.ResourcesDir)/default/netcore/bin/
 ```
 
-### Publish a named artifact
+### Upload a named artifact
 
-You can give an artifact a name, and you can publish multiple named artifacts. All artifacts are downloaded unless you specify a `downloadArtifact` step to limit the artifacts that are downloaded.
+You can give an artifact a name, and you can upload multiple named artifacts. All artifacts are downloaded unless you specify a `download` step to limit the artifacts that are downloaded.
 
 ```yaml
 - job: Build
   steps:
   - script: dotnet publish --configuration $(buildConfiguration)
-  - artifact: bin/WebApp/*
+  - upload: bin/WebApp/*
     name: WebApp
-  - artifact: bin/MobileApp/*
+  - upload: bin/MobileApp/*
     name: MobileApp
 - job: Deploy
   steps:
-  - downloadArtifact: WebApp
-  - downloadArtifact: MobileApp
+  - download: WebApp
```

Comment:
In this example WebApp and MobileApp are pipeline artifacts and not resources. So the schema has to be 
-download: 
  name: WebApp
-download: 
  name: MobileApp

Reply:
There is a possibility that the internal artifact name "WebApp" can collide with pipeline resource name. 
And one input type is supposed to take one type of object. In this case `-download` will end up having two possible input types (pipeline resource or internal artifact name). This will be confusing. Alex and I discussed about this when he was in IDC.

Reply:
Is the risk greater here than it was with containers? Here are two legal YAML files which do something quite different:

```yaml
# explicit resource
resources:
  containers:
  - container: python
    image: python:2.7-windowsservercore-ltsc2016

container: python  # a resource pointing to the Python 2.7 container on Windows
steps:
- script: echo foo
```

```yaml
# no explicit resource
container: python  # no resource found; this is the default tag for Python, probably a 3.7.x on Linux
steps:
- script: echo foo
```

But in both cases, these files do (likely) exactly what the user expects.

Reply:
Yeah, I realize we follow the same pattern for containers.

With templates, there is very good possibility that build stage where artifacts are published can be defined in some other file. And user wouldn't know what is the artifact name given while publishing unless he/she verifies all the files carefully. 

Moreover, this also means there are two different ways of doing the same thing:
-download:
  name: WebTier
vs
-download: WebTier

And the same input field taking two different types of objects (resource and artifact) can also be confusing.

Reply:
OK, I think I conflated two things. After reading more carefully, I agree with @ashokirla here.